yt_dlp/networking/common.py

```diff
@@ -0,0 +1,431 @@
+        self._set_ydl(ydl)
+        self.cookiejar = self.ydl.cookiejar
+
+    def _set_ydl(self, ydl):
+        # TODO: this sucks
+        self.ydl = ydl
+
+        for func in (
+            'deprecation_warning',
+            'report_warning',
+            'to_stderr',
+            'write_debug',
+            'to_debugtraffic'
+        ):
+            if not hasattr(self, func):
+                setattr(self, func, getattr(ydl, func))
+
+    def make_sslcontext(self, **kwargs):
+        """
+        Make a new SSLContext configured for this backend.
+        Note: _make_sslcontext must be implemented
+        """
+        context = self._make_sslcontext(
+            verify=not self.ydl.params.get('nocheckcertificate'), **kwargs)
+        if not context:
+            return context
+        if self.ydl.params.get('legacyserverconnect'):
+            context.options |= 4  # SSL_OP_LEGACY_SERVER_CONNECT
+            # Allow use of weaker ciphers in Python 3.10+. See https://bugs.python.org/issue43998
+            context.set_ciphers('DEFAULT')
+
+        client_certfile = self.ydl.params.get('client_certificate')
+        if client_certfile:
+            try:
+                context.load_cert_chain(
+                    client_certfile, keyfile=self.ydl.params.get('client_certificate_key'),
+                    password=self.ydl.params.get('client_certificate_password'))
+            except ssl.SSLError:
+                raise YoutubeDLError('Unable to load client certificate')
+
+        return context
+
+    def _make_sslcontext(self, verify: bool, **kwargs) -> ssl.SSLContext:
+        """Generate a backend-specific SSLContext. Redefine in subclasses"""
+        raise NotImplementedError
+
+    def prepare_request(self, request: Request):
+        super().prepare_request(request)
+        request.headers = CaseInsensitiveDict(self.ydl.params.get('http_headers', {}), request.headers)
+        if request.headers.get('Youtubedl-no-compression'):
+            request.compression = False
+            del request.headers['Youtubedl-no-compression']
+
+        # Proxy preference: header req proxy > req proxies > ydl opt proxies > env proxies
+        request.proxies = {**(self.ydl.proxies or {}), **(request.proxies or {})}
+        req_proxy = request.headers.get('Ytdl-request-proxy')
+        if req_proxy:
+            del request.headers['Ytdl-request-proxy']
+            request.proxies.update({'http': req_proxy, 'https': req_proxy})
+        for proxy_key, proxy_url in request.proxies.items():
+            if proxy_url == '__noproxy__':  # compat
+                request.proxies[proxy_key] = None
+                continue
+
+            if proxy_url is not None and _parse_proxy is not None:
+                # Ensure proxies without a scheme are http.
+                proxy_scheme = _parse_proxy(proxy_url)[0]
+                if proxy_scheme is None:
+                    request.proxies[proxy_key] = 'http://' + remove_start(proxy_url, '//')
```

Comment:
Some libraries may need specific proxy sanitization before you pass the proxies to them.

Also `select_proxy` may not be used in a backend (which is fine as long as they implement similar behavior).

It is only really needed for `urllib` at this stage. I use it for requests since it's easy to implement, but the behaviour is pretty much the same. Requests by default supports more complex proxy dict key definitions, which this overrides, though I don't think it is something we have to worry about (unless someone is passing `proxies` to `Request).